src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -311,126 +401,58 @@ impl Net {
+            || self.has_feature(VIRTIO_NET_F_GUEST_UFO as u64)
+        {
+            65562
+        } else {
+            1526
         }
-
-        warn!("Receiving buffer is too small to hold frame of current size");
-        Err(FrontendError::DescriptorChainTooSmall)
     }
 
-    // Copies a single frame from `self.rx_frame_buf` into the guest.
-    fn do_write_frame_to_guest(&mut self) -> Result<(), FrontendError> {
+    /// Parse available RX `DescriptorChains` from the queue
+    pub fn parse_rx_descriptors(&mut self) {
         // This is safe since we checked in the event handler that the device is activated.
         let mem = self.device_state.mem().unwrap();
-
         let queue = &mut self.queues[RX_INDEX];
-        let head_descriptor = queue.pop_or_enable_notification().ok_or_else(|| {
-            self.metrics.no_rx_avail_buffer.inc();
-            FrontendError::EmptyQueue
-        })?;
-        let head_index = head_descriptor.index;
-
-        let result = Self::write_to_descriptor_chain(
-            mem,
-            &self.rx_frame_buf[..self.rx_bytes_read],
-            head_descriptor,
-            &self.metrics,
-        );
-        // Mark the descriptor chain as used. If an error occurred, skip the descriptor chain.
-        let used_len = if result.is_err() {
-            self.metrics.rx_fails.inc();
-            0
-        } else {
-            // Safe to unwrap because a frame must be smaller than 2^16 bytes.
-            u32::try_from(self.rx_bytes_read).unwrap()
-        };
-        queue.add_used(head_index, used_len).map_err(|err| {
-            error!("Failed to add available descriptor {}: {}", head_index, err);
-            FrontendError::AddUsed
-        })?;
-
-        result
-    }
-
-    // Copies a single frame from `self.rx_frame_buf` into the guest. In case of an error retries
-    // the operation if possible. Returns true if the operation was successfull.
-    fn write_frame_to_guest(&mut self) -> bool {
-        let max_iterations = self.queues[RX_INDEX].actual_size();
-        for _ in 0..max_iterations {
-            match self.do_write_frame_to_guest() {
-                Ok(()) => return true,
-                Err(FrontendError::EmptyQueue) | Err(FrontendError::AddUsed) => {
-                    return false;
+        while let Some(head) = queue.pop_or_enable_notification() {
+            let index = head.index;
+            // SAFETY: we are only using this `DescriptorChain` here.
+            if let Err(err) = unsafe { self.rx_buffer.add_buffer(mem, head) } {
+                // If guest uses dirty tricks to make us add more descriptors than
+                // we can hold, just stop processing.
+                if matches!(err, AddRxBufferError::Parsing(IoVecError::IovDequeOverflow)) {
+                    error!("net: Could not add an RX descriptor: {err}");
+                    queue.undo_pop();
+                    break;
                 }
```

Comment:
I think we should add this check after increasing the metric below and adding back the descriptor chain as used.

Reply:
> and adding back the descriptor chain as used

Reply:
Also `rx_fails` is a horrible metric name for this. We need to update this at some point.